File: haskell/src/Surd/Internal/PrimeFactors.hs
-- | Small prime factorization and related utilities.
module Surd.Internal.PrimeFactors
  ( factorise
  , primeFactors
  , isPrime
  , primes
  ) where

import Data.List.NonEmpty (NonEmpty(..))
import qualified Data.List.NonEmpty as NE
import Surd.Internal.Positive (Positive, unPositive)

-- | Factorise a positive integer into (prime, exponent) pairs.
--
-- >>> factorise 360
-- [(2,3),(3,2),(5,1)]
factorise :: Positive -> [(Integer, Int)]
factorise pos =
  let n = toInteger (unPositive pos)
  in if n == 1
     then []
     else map (\(x :| xs) -> (x, 1 + length xs)) . NE.group $ go n primes
  where
    go 1 _      = []
    go m (p:ps)
      | p * p > m = [m]
      | r == 0    = p : go q (p:ps)
      | otherwise = go m ps
      where (q, r) = m `divMod` p
    go _ []     = error "primes exhausted"  -- unreachable

-- | Distinct prime factors of a positive integer, in ascending order.
--
-- >>> primeFactors 360
-- [2,3,5]
primeFactors :: Positive -> [Integer]
primeFactors = map fst . factorise

-- | Trial-division primality test (adequate for the sizes we encounter).
isPrime :: Integer -> Bool
isPrime n
  | n < 2     = False
  | n < 4     = True
  | otherwise = all (\p -> n `mod` p /= 0) $ takeWhile (\p -> p * p <= n) primes

-- | Infinite list of primes via trial division.
primes :: [Integer]
primes = 2 : filter isPrime [3, 5 ..]

File: scala3/src/main/scala/surd/PrimeFactors.scala
package surd

/** Prime factorization and related utilities via trial division. */
object PrimeFactors:

  /** Infinite lazy stream of primes via trial division. */
  lazy val primes: LazyList[BigInt] =
    BigInt(2) #:: LazyList.from(3, 2).map(BigInt(_)).filter(isPrime)

  /** Trial-division primality test (adequate for moderate-sized integers). */
  def isPrime(n: BigInt): Boolean =
    if n < 2 then false
    else if n < 4 then true
    else primes.takeWhile(p => p * p <= n).forall(p => n % p != 0)

  def isPrime(n: Int): Boolean = isPrime(BigInt(n))

  /** Factorise a positive integer into (prime, exponent) pairs in ascending order.
    *
    * {{{
    * scala> factorise(Positive.unsafe(360))
    * Vector((2,3), (3,2), (5,1))
    * }}}
    */
  def factorise(pos: Positive): Vector[(BigInt, Int)] =
    val n = pos.value
    if n == 1 then Vector.empty
    else
      val factors = trialDivide(n, primes)
      // Group consecutive equal primes and count
      groupConsecutive(factors)

  private def trialDivide(n: BigInt, ps: LazyList[BigInt]): Vector[BigInt] =
    if n == 1 then Vector.empty
    else
      ps match
        case p #:: rest =>
          if p * p > n then Vector(n)
          else if n % p == 0 then p +: trialDivide(n / p, ps)
          else trialDivide(n, rest)
        case _ => Vector.empty // unreachable

  /** Group a sorted vector of values into (value, count) pairs. */
  private def groupConsecutive(vs: Vector[BigInt]): Vector[(BigInt, Int)] =
    if vs.isEmpty then Vector.empty
    else
      val builder = Vector.newBuilder[(BigInt, Int)]
      var current = vs.head
      var count = 1
      for v <- vs.tail do
        if v == current then count += 1
        else
          builder += ((current, count))
          current = v
          count = 1
      builder += ((current, count))
      builder.result()

  /** Distinct prime factors in ascending order. */
  def primeFactors(pos: Positive): Vector[BigInt] =
    factorise(pos).map(_._1)
